### PR TITLE
Draft: NT4.1 support

### PR DESCRIPTION
**Pull request by CrispyBacon1999**
## Description

This pull request introduces support for NetworkTables v4.1 while maintaining backward compatibility with version 4.0. The primary goal is to enable seamless integration with NetworkTables servers that have upgraded to v4.1, while ensuring that existing functionality for v4.0 remains intact.

Currently, there are no tests written for 4.1, which should absolutely be added. All the tests for v4.0 still pass, with minor modifications to ensure that the code is running at 4.0 instead.

## Changes Made

New Feature: Added support for NetworkTables version 4.1.

Compatibility: The code has been structured to preserve compatibility with NetworkTables version 4.0. The default behavior remains as it was for version 4.0, ensuring that users running older servers won't experience any disruptions.

Conditional Logic: Implemented conditional logic to detect the server's version during runtime. If the server supports version 4.1, the code will automatically switch to using the new features and improvements introduced in this version.

## Questions for Reviewers
Are there any concerns about maintaining backward compatibility with version 4.0? Have we covered all possible scenarios where compatibility might be affected?

Have we implemented the conditional logic for version detection in an efficient and reliable manner? Are there any potential edge cases that we might have missed?

Are there any performance considerations or optimizations that need to be addressed to ensure the smooth operation of the code with version 4.1?

---

Any feedback and suggestions for improvement are highly welcomed.

Note: The changes for the documentation make PRs get a bit messy. Would it be worthwhile changing it so the docs build only on merges to main?

### CODE REVIEW

File: packages/ntcore-ts-client/src/lib/socket/socket.ts [hunk 1]
```diff
@@ -25,6 +25,12 @@ export class NetworkTablesSocket {
   private offset = 0;
   private bestRtt = -1;
 
+  private RECONNECT_TIMEOUT = 1000;
+  private RTT_PERIOD_V4_0 = 1000;
+  private RTT_PERIOD_V4_1 = 250;
+  private TIMEOUT_V4_0 = 3000;
```
Comment (CrispyBacon1999) [inline packages/ntcore-ts-client/src/lib/socket/socket.ts:31]: This is temporarily unused, as v4.0 doesn't have the same timeout logic. If we wanted to add it, 3000ms is probably an ideal amount.

File: packages/ntcore-ts-client/src/lib/socket/socket.ts [hunk 2]
```diff
@@ -124,12 +135,29 @@ export class NetworkTablesSocket {
...
         this.sendQueuedMessages();
 
+        if (this._websocket.protocol === 'v4.1.networktables.first.wpi.edu') {
+          // eslint-disable-next-line no-console
+          console.info('Connected on NT 4.1');
+          this.initializeRTTWebsocket();
+        } else {
+          // eslint-disable-next-line no-console
+          console.info('Connected on NT 4.0');
+        }
```
Comment (CrispyBacon1999) [inline packages/ntcore-ts-client/src/lib/socket/socket.ts:145]: Not sure if we need to broadcast this to users, it's very helpful for debugging though.

File: packages/ntcore-ts-client/src/lib/socket/socket.ts [hunk 3]
```diff
@@ -160,6 +191,35 @@ export class NetworkTablesSocket {
...
 
+  private initializeRTTWebsocket() {
+    if (this._rttWebsocket === null) {
+      this._rttWebsocket = new WebSocket(this.serverUrl, 'rtt.networktables.first.wpi.edu');
+      this._rttWebsocket.binaryType = 'arraybuffer';
+      this.resetTimeout();
+      this._rttWebsocket.onmessage = (event: MessageEvent | WS_MessageEvent) => this.onMessage(event, true);
+      this._rttWebsocket.onerror = (event: Event | WS_ErrorEvent) => this.onError(event);
+    }
+  }
```
Comment (CrispyBacon1999) [inline packages/ntcore-ts-client/src/lib/socket/socket.ts:202]: This needs documentation still.

File: packages/ntcore-ts-client/src/lib/socket/socket.ts [hunk 4]
```diff
@@ -168,7 +228,10 @@ export class NetworkTablesSocket {
   reinstantiate(serverUrl: string) {
     this.close();
     this.serverUrl = serverUrl;
-    this._websocket = new WebSocket(this.serverUrl, 'networktables.first.wpi.edu');
+    this._websocket = new WebSocket(this.serverUrl, [
+      'v4.1.networktables.first.wpi.edu',
+      'networktables.first.wpi.edu',
+    ]);
```
Comment (CrispyBacon1999) [inline packages/ntcore-ts-client/src/lib/socket/socket.ts:234]: This should probably be abstracted to a function? Particularly with the protocol array, it gets recreated several times throughout this file and will likely become problematic in the future when v4.2 inevitably comes out.

**Comment by cjlawson02** [comment]
Hi, please merge in beta and run CI again. I took a quick glance... did we implement the PING and PONG requirements? Maybe RTT sufices

**Comment by CrispyBacon1999** [comment]
Not 100% sure what's needed for the PING and PONG. Most of the logic for this was pulled from AdvantageScope since they had already done the upgrade and have their own library for it.

**Comment by cjlawson02** [comment]
hi @CrispyBacon1999 can you rebase? I got it working with 4.1 on the beta branch, but your code might have improvements

**Comment by cjlawson02** [comment]
@CrispyBacon1999 If you have hardware to test this on, I rebased in beta

**Comment by CrispyBacon1999** [comment]
I don't have it tonight, but I should be able to test tomorrow if I remember to.

**Comment by CrispyBacon1999** [comment]
I was able to test the connection to a physical robot tonight and it worked perfectly! Connected properly over NT4.1 and I was able to poll values from the robot.

**Review by cjlawson02 — APPROVED** [review]

**Pull request merged**